### PR TITLE
feat: add memlimit flags for setMaxResources

### LINKED ISSUE #18849: feat: Correctly setting GOMEMLIMIT to prevent OOM generation
**Is your feature request related to a problem? Please describe.**
Correctly setting GOMEMLIMIT to prevent OOM generation 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jiuker**
feat: add memlimit flags for runtime

## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description

add memlimit flags for setMaxResources
fix https://github.com/minio/minio/issues/18849

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/server-rlimit.go [hunk 1]
```diff
@@ -75,6 +77,25 @@ func setMaxResources() (err error) {
...
+			memlimit = ctx.GlobalString("memlimit")
+		}
+		mlimit, err := humanize.ParseBytes(memlimit)
+		if err != nil {
+			return err
+		}
+		debug.SetMemoryLimit(int64(mlimit))
+		return nil
+	}
+	return nil
```
Comment (shtripat) [inline cmd/server-rlimit.go]:
```suggestion
        if ctx == nil {
            return nil
        }
	if ctx.IsSet("memlimit") {
		memlimit := ctx.String("memlimit")
		if memlimit == "" {
			memlimit = ctx.GlobalString("memlimit")
		}
		mlimit, err := humanize.ParseBytes(memlimit)
		if err != nil {
			return err
		}
		debug.SetMemoryLimit(int64(mlimit))
	}
	return nil
```

**Review by shtripat — COMMENTED** [review]
Mostly looks good. A simple suggestion.

**Review by shtripat — APPROVED** [review]
lgtm

File: cmd/server-rlimit.go [hunk 2]
```diff
@@ -75,6 +77,27 @@ func setMaxResources() (err error) {
...
+
 	err = sys.SetMaxMemoryLimit(maxLimit, maxLimit)
 	return err
 }
+
+func setDebugMemoryLimit(ctx *cli.Context) error {
+	if ctx == nil {
+		return nil
+	}
+	if ctx.IsSet("memlimit") {
```
Comment (vadmeste) [inline cmd/server-rlimit.go:91]: I do not see memlimit flag defined anywhere, you need to define it ServerFlags in server-main.go

**Review by vadmeste — CHANGES_REQUESTED** [review]

File: cmd/server-main.go [hunk 3]
```diff
@@ -172,6 +172,11 @@ var ServerFlags = []cli.Flag{
 		Hidden: true,
 		EnvVar: "MINIO_CROSSDOMAIN_XML",
 	},
+	cli.StringFlag{
+		Name:   "memlimit",
+		Usage:  "set debug memory limit instead of GOMEMLIMIT env",
```
Comment (harshavardhana) [inline cmd/server-main.go]:
```suggestion
		Usage:  "set global memory limit per server via GOMEMLIMIT",
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

File: cmd/server-main.go [hunk 4]
```diff
@@ -172,6 +172,11 @@ var ServerFlags = []cli.Flag{
 		Hidden: true,
 		EnvVar: "MINIO_CROSSDOMAIN_XML",
 	},
+	cli.StringFlag{
+		Name:   "memlimit",
+		Usage:  "set global memory limit per server via GOMEMLIMIT",
+		Hidden: true,
```
Comment (harshavardhana) [inline cmd/server-main.go:178]:
```suggestion
		Hidden: true,
		EnvVar: "MINIO_MEMLIMIT",
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**